contracts/mirror/src/contract.rs

```diff
@@ -395,72 +304,90 @@ pub fn sudo(
+    // for given user our contract owns these failed-to-process tokens
+    FAILED_TRANSFERS.update(
+        deps.storage,
+        packet.receiver,
+        |current_debt| match current_debt {
+            Some(funds) => Ok::<Vec<FailedCoin>, ContractError>(
+                // If given denom exist - just modify its amount and attempt
+                if funds.iter().any(|coin| coin.coin.denom == packet.denom) {
+                    funds
+                        .iter()
+                        .map(|coin| {
+                            if coin.coin.denom == packet.denom {
+                                FailedCoin {
+                                    // coin.attempt should be summed up with memo here because of the following situation:
+                                    // 1) We have failed balance of {attempt: 13, coin {denom: denom, amount: 123}} |-----------*
+                                    // 2) We execute retry, this value then removed from the map to prevent retry attack        |
+                                    // 3) While operation above is asynchronous, we're doing execute_bond, it fails             |
+                                    //    and it adds a new value into the map {attempt: 0, coin {denom: denom, amount: 321}}   |
+                                    // 4) sudo handler, initiated by the 1-st operation, executes this piece of code  <---------*
+                                    //    and in this case attempt should be taken from the memo and then incremented.
+                                    //    Also, value from the 3-rd clause could be retried also several times and it's reason why we
+                                    //    should sum up the memo and current attempt.
+                                    attempt: coin.attempt + memo_attempt + 1,
+                                    coin: Coin {
+                                        denom: coin.coin.denom.clone(),
+                                        amount: coin.coin.amount + packet_amount,
+                                    },
+                                }
+                            } else {
+                                coin.clone()
+                            }
+                        })
+                        .collect()
+                // If it doesn't exist - push it at the end
+                } else {
+                    funds
+                        .iter()
+                        .cloned()
+                        .chain(std::iter::once(FailedCoin {
+                            attempt: memo_attempt + 1,
+                            coin: Coin {
+                                denom: packet.denom.clone(),
+                                amount: packet_amount,
+                            },
+                        }))
+                        .collect()
+                },
+            ),
+            None => Ok(vec![FailedCoin {
+                attempt: memo_attempt + 1,
```

Comment:
Why do you increment memo twice? Here and on line 320. Also in a different code path on line 346.